clippy_lints/src/hash_borrow_str_semantics.rs

```diff
@@ -0,0 +1,114 @@
+    /// type that implements all three of Hash, Borrow<str> and Borrow<[u8]>
+    /// as it is impossible to satisfy the semantics of Borrow and Hash for
+    /// both Borrow<str> and Borrow<[u8]>.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// When providing implementations for Borrow<T>, one should consider whether the different
+    /// implementations should act as facets or representations of the underlying type. Generic code
+    /// typically uses Borrow<T> when it relies on the identical behavior of these additional trait
+    /// implementations. These traits will likely appear as additional trait bounds.
+    ///
+    /// In particular Eq, Ord and Hash must be equivalent for borrowed and owned values:
+    /// `x.borrow() == y.borrow()` should give the same result as `x == y`.
+    /// It follows then that the following equivalence must hold:
+    /// `hash(x) == hash((x as Borrow<[u8]>).borrow()) == hash((x as Borrow<str>).borrow())`
+    ///
+    /// Unfortunately it doesn't hold as `hash("abc") != hash("abc".as_bytes())`.
+    ///
+    /// ### Example
+    ///
+    /// ```
+    /// use std::borrow::Borrow;
+    /// use std::hash::{Hash, Hasher};
+    ///
+    /// struct ExampleType {
+    ///     data: String
+    /// }
+    ///
+    /// impl Hash for ExampleType {
+    ///     fn hash<H: Hasher>(&self, state: &mut H) {
+    ///         self.data.hash(state);
+    ///     }
+    /// }
+    ///
+    /// impl Borrow<str> for ExampleType {
+    ///     fn borrow(&self) -> &str {
+    ///         &self.data
+    ///     }
+    /// }
+    ///
+    /// impl Borrow<[u8]> for ExampleType {
+    ///     fn borrow(&self) -> &[u8] {
+    ///         self.data.as_bytes()
+    ///     }
+    /// }
+    ///
+    /// fn hash(x: impl Hash) -> u64 {
+    ///    let mut hasher = std::collections::hash_map::DefaultHasher::new();
+    ///    x.hash(&mut hasher);
+    ///    hasher.finish()
+    /// }
+    ///
+    /// fn main() {
+    ///    let example = ExampleType { data: "abc".to_string() };
+    ///    assert_eq!(hash(&example), hash(example.borrow() as &str));
+    ///    assert_ne!(hash(&example), hash(example.borrow() as &[u8]));
+    /// }
+    ///
+    /// ```
+    ///
+    #[clippy::version = "1.75.0"]
+    pub HASH_BORROW_STR_SEMANTICS,
+    correctness,
+    "Ensures that the semantics of Borrow for Borrow<str> and Borrow<[u8]> are satisfied for Hash"
+}
+
+declare_lint_pass!(HashBorrowStrSemantics => [HASH_BORROW_STR_SEMANTICS]);
+
+impl LateLintPass<'_> for HashBorrowStrSemantics {
+    /// We are emitting this lint at the Hash impl of a type that implements all
+    /// three of Hash, Borrow<str> and Borrow<[u8]>.
+    ///
+    /// We check that we are in the Hash impl
+    /// Then we check that the type implements Borrow<str> and Borrow<[u8]>
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        if let ItemKind::Impl(imp) = item.kind &&
```

Comment:
Nit (I'm surprised rustfmt didn't pick this up?): The `&&` should normally go on a new line. See e.g. here

https://github.com/rust-lang/rust-clippy/blob/d487579efd3c6b7766f820bbe4b6be0f6d420dad/clippy_lints/src/attrs.rs#L661-L663